### PR TITLE
Revamp profile handling CLI flags

### PR DESCRIPTION
**Pull request by fgsch**
This PR moves profiling to `--profile`, and replaces `--profile-guest=[path]` in favour of `--profile=guest[,path]`, unifying the profiling options (and matching Wasmtime). Most of the code is adapted from Wasmtime.

While here I've removed some unncessary lifetimes and polished the code, mostly courtesy of clippy (the 2 last commits).

The PR is related to and was built on top of https://github.com/fastly/Viceroy/pull/316. It might make sense to review that one first.

### CODE REVIEW

File: lib/src/wiggle_abi/req_impl.rs [hunk 1]
```diff
@@ -95,9 +95,9 @@ impl FastlyHttpReq for Session {
     }
 
     #[allow(unused_variables)] // FIXME JDC 2023-06-18: Remove this directive once implemented.
-    fn downstream_client_h2_fingerprint<'a>(
```
Comment (JakeChampion) [inline lib/src/wiggle_abi/req_impl.rs:98]: The changes in this file don't look to be related to the description of this pull-request - could we have these in a different pull-request please? 🤞

File: cli/src/opts.rs [hunk 2]
```diff
@@ -48,11 +48,6 @@ pub struct ServeArgs {
     #[arg(long = "addr")]
     socket_addr: Option<SocketAddr>,
 
-    /// Whether to profile the wasm guest. Takes an optional directory to save
-    /// per-request profiles to
-    #[arg(long, default_missing_value = "guest-profiles", num_args=0..=1, require_equals=true)]
-    profile_guest: Option<PathBuf>,
-
```
Comment (JakeChampion) [inline cli/src/opts.rs:55]: Note: This would be a breaking change as we are removing a flag, we'd need to version accordingly

File: lib/src/execute.rs [hunk 3]
```diff
@@ -84,7 +84,7 @@ impl ExecuteCtx {
         module_path: impl AsRef<Path>,
         profiling_strategy: ProfilingStrategy,
         wasi_modules: HashSet<ExperimentalModule>,
-        guest_profile_path: Option<PathBuf>,
+        guest_profile_path: Option<String>,
```
Comment (JakeChampion) [inline lib/src/execute.rs:87]: I think we can avoid this breaking change in the viceroy_lib crate by applying this change instead:
```diff
diff --git a/cli/src/main.rs b/cli/src/main.rs
index bcf7524..db46cdf 100644
--- a/cli/src/main.rs
+++ b/cli/src/main.rs
@@ -240,7 +240,7 @@ async fn create_execution_context(
         input,
         args.profiling_strategy(),
         args.wasi_modules(),
-        guest_profile_path,
+        guest_profile_path.map(|p| p.into()),
         args.unknown_import_behavior(),
     )?
     .with_log_stderr(args.log_stderr())
diff --git a/lib/src/execute.rs b/lib/src/execute.rs
index bb45d86..df731fe 100644
--- a/lib/src/execute.rs
+++ b/lib/src/execute.rs
@@ -84,7 +84,7 @@ impl ExecuteCtx {
         module_path: impl AsRef<Path>,
         profiling_strategy: ProfilingStrategy,
         wasi_modules: HashSet<ExperimentalModule>,
-        guest_profile_path: Option<String>,
+        guest_profile_path: Option<PathBuf>,
         unknown_import_behavior: UnknownImportBehavior,
     ) -> Result<Self, Error> {
         let config = &configure_wasmtime(profiling_strategy);
@@ -112,7 +112,6 @@ impl ExecuteCtx {
                 engine_clone.increment_epoch();
             }
         })));
-        let guest_profile_path = guest_profile_path.map(PathBuf::from);
 
         Ok(Self {
             engine,
```

Comment (fgsch) [reply lib/src/execute.rs]: I've updated this PR to deal with the breaking change.

**Review by JakeChampion — APPROVED** [review]
